components/omega/src/ocn/HorzMesh.h

```diff
@@ -0,0 +1,192 @@
+#ifndef OMEGA_HORZMESH_H
+#define OMEGA_HORZMESH_H
+//===-- base/HorzMesh.h - horizontal mesh --------------------*- C++ -*-===//
+//
+/// \file
+/// \brief Contains the mesh variables for an OMEGA horizontal sub-domain
+///
+/// The HorzMesh class contains the data to represent a sub-domain of the 
+/// global horizontal mesh.  
+//
+//===----------------------------------------------------------------------===//
+
+#include "DataTypes.h"
+#include "MachEnv.h"
+
+#include <string>
+
+namespace OMEGA {
+
+/// A class for the horizontal mesh information
+
+/// The HorzMesh class reads in the remaining mesh information that is
+/// not used to create the domain decomposition from mesh file.
+/// It handles computing any dependent mesh quantities and transfers 
+/// the relevant information to the device
+class HorzMesh {
+
+ private:
+
+    void initParallelIO(Decomp *MeshDecomp);
+
+    void readCoordinates();
+
+    void readBottomDepth();
+
+    void readMeasurements();
+
+    void readWeights();
+
+    void readCoriolis();
+
+    void copyToDevice();
+
+    //int computeMesh();
+
+    I4 CellDecompR8;
+    I4 EdgeDecompR8;
+    I4 VertexDecompR8;
+    I4 OnEdgeDecompR8;
+    I4 OnVertexDecompR8;
+
+ public:
+   // Variables
+   // Since these are used frequently, we make them public to reduce the
+   // number of retrievals required.
+
+   std::string MeshFileName;
+   int MeshFileID;
+
+   // Sizes and global IDs
+   // Note that all sizes are actual counts (1-based) so that loop extents
+   // should always use the 0:NCellsXX-1 form.
+
+   I4 NCellsOwned;  ///< Number of cells owned by this task
+   I4 NCellsAll;    ///< Total number of local cells (owned + all halo)
+   I4 NCellsSize;   ///< Array size (incl padding, bndy cell) for cell arrays
+
+   I4 NEdgesOwned;    ///< Number of edges owned by this task
+   I4 NEdgesAll;      ///< Total number (owned+halo) of local edges
+   I4 NEdgesSize;     ///< Array length (incl padding, bndy) for edge dim
+   I4 MaxCellsOnEdge; ///< Max number of cells sharing an edge
+   I4 MaxEdges;       ///< Max number of edges around a cell
+   I4 MaxEdges2;      ///< Max number of edges around a cell x2
+
+   I4 NVerticesOwned;  ///< Number of vertices owned by this task
+   I4 NVerticesAll;    ///< Total number (owned+halo) of local vertices
+   I4 NVerticesSize;   ///< Array length (incl padding, bndy) for vrtx dim
+   I4 VertexDegree;    ///< Number of cells that meet at each vertex
+
+   // Mesh connectivity
+
+   Array2DI4 CellsOnCell;      ///< Indx of cells that neighbor each cell
+   ArrayHost2DI4 CellsOnCellH; ///< Indx of cells that neighbor each cell
```

Comment:
This pattern of declaring two arrays (host and device) per mesh variable will double the amount of memory needed to store the mesh when the device is a CPU. Additionally, redundant copies between the two arrays will be performed. It is probably fine for now, but long-term we might need a different solution.